### PR TITLE
301 undo

### LINKED ISSUE #301: Undo geometry changes
- [x] Add undo button
- [x] On undo -> (creating routepath geometry) take last modification from modification store and apply it to the item store

### PR DESCRIPTION
**Pull request by kronken**
Closes #301 

TO test: Start creating new routepath geometry, you should now be able to undo your geometry changes using the undo button in the toolbar.

### CODE REVIEW

File: src/enums/eventType.ts [hunk 1]
```diff
@@ -1,7 +1,5 @@
 enum eventType {
-    ADD= 'Lisäys',
-    DELETE= 'Poisto',
-    MOVE= 'Siirto',
+    ADD_ROUTEPATH_LINK = 'Reitinlinkin lisäys',
```
Comment (kronken) [inline src/enums/eventType.ts]: Open for suggestions for this event in finnish.

File: src/stores/geometryEventStore.ts [hunk 2]
```diff
@@ -68,8 +92,25 @@ export class GeometryEventStore {
...
+        if (routePathLinks.length > 0) {
+            const neighbourLinks =
+                await RoutePathLinkService.fetchAndCreateRoutePathLinksWithStartNodeId(
+                    routePathLinks[routePathLinks.length - 1].endNode.id,
+                );
+            RoutePathStore!.setNeighborRoutePathLinks(neighbourLinks);
+        } else {
+            RoutePathStore!.setNeighborRoutePathLinks([]);
+        }
+    }
```
Comment (Divinare) [inline src/stores/geometryEventStore.ts]: This method doens't seem belong to geometryEventStore. GeometryEventStore should be as generic as possible.

I know this was an old decision but I would put everything related routePaths somewhere else. What the whole logic was moved to newRoutePathLayer? I think its newRoutePath layer's responsibility to query neighborLinks when it needs

File: src/stores/geometryEventStore.ts [hunk 3]
```diff
@@ -31,29 +33,51 @@ export class GeometryEventStore {
...
+            entityName: entityName,
+        },
+        ) {
+        this._events.push(
+            {
+                action,
+                preObject,
+                timestamp: new Date(),
+                entity: entityName,
+            });
```
[the remark this answers is not in the document]
Comment (Divinare) [reply src/stores/geometryEventStore.ts]: a = routePath: { links: [123] }
b = routePath: { links: [123, 444] }
c = routePath: { links: [123, 444, 551 }

now our geometryEventStore would look like:
[a, b, c]

when I undo to [a, b], I call routePathStore.setRoutePath(b);

**Comment by Divinare** [comment]
Testing was ok

File: src/components/map/toolbar/ToolbarCommonButtons.tsx [hunk 4]
```diff
@@ -29,17 +36,22 @@ class ToolbarCommonButtons extends React.Component {
...
-                        </a>
+                        <FiExternalLink />
+                    </MapControlButton>
+                </div>
+                <div className={s.toolbarButtonRow}>
+                    <MapControlButton
+                        onClick={this.undo}
+                        isActive={false}
+                        isDisabled={!GeometryEventStore.hasEvents}
+                        label='Undo'
```
[the remark this answers is not in the document]
Comment (kronken) [reply src/components/map/toolbar/ToolbarCommonButtons.tsx]: Yes, my initial intention was to add CTRL + z, but that was more complicated than what I thought. I didn't want to make this task too big, so better that it is in a new task

File: src/stores/geometryEventStore.ts [hunk 5]
```diff
@@ -31,29 +33,51 @@ export class GeometryEventStore {
...
     }
 
-    private logRoutePathLinkChanges(change: IObjectDidChange) {
+    @action
+    public undo() {
+        const event = this._events.pop();
+        if (event) {
+            switch (event.action) {
+            case eventType.ADD_ROUTEPATH_LINK:
```
[the remark this answers is not in the document]
Comment (kronken) [reply src/stores/geometryEventStore.ts]: So, if we don't know what action it is. How do we know what the user has modified?

File: src/stores/geometryEventStore.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (kronken) [reply src/stores/geometryEventStore.ts]: You last row, how to you know to call routePathStore.setRoutePath(b)? How do you know that the last geometry change has anything to do with routePaths?

File: src/stores/geometryEventStore.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (kronken) [reply src/stores/geometryEventStore.ts]: Once again, same as other functions, how do you know which store to call if you don't know what kind of object it is?

File: src/stores/geometryEventStore.ts [hunk 6]
```diff
@@ -31,29 +33,51 @@ export class GeometryEventStore {
...
-                .forEach((routePathLink: IRoutePathLink) => {
+                .forEach(() => {
+                    const preRoutePathList = routePathLinks.slice();
+                    preRoutePathList.splice(change['index'], 1);
                     this.pushToEvents({
-                        action: eventType.ADD,
+                        action: eventType.ADD_ROUTEPATH_LINK,
                         entityName: entityName.ROUTELINK,
-                        objectId: routePathLink.id,
+                        preObject: preRoutePathList,
```
[the remark this answers is not in the document]
Comment (kronken) [reply src/stores/geometryEventStore.ts]: typeof [] = 'object'
Actually Javascript arrays are objects.
But that is not the point of this thing, we have a generic previous item/object/value/whatever. Do you have a suggestion that would be more logical?

File: src/stores/geometryEventStore.ts (hunk 2 above)
Comment (kronken) [reply src/stores/geometryEventStore.ts]: The first implementation of this had this logic in the component, then we decided we wanted it isolated from the component. So, we refactored all the logic, and isolated it from the components.

Of course we can't have this function in this file forever. I see this as a first step, and when we get another function extracting the logic to some new layer.

**Pull request merged**